### PR TITLE
Clean-up the HTTP versions check in response parsing, and don't reuse connections on any response other than HTTP/1.1

### PR DESCRIPTION
**Pull request by ahsonkhan**
**Update:** Other than making the HTTP version checks a bit easier to understand (using single conditional if, as suggested by @LarryOsterman), the **primary** change here is to not persist a connection if we received an invalid/unsupported HTTP version in the response, i.e. if the HTTP version does not start with 1.x. We could throw here instead, but at least marking it as don't re-use, is a good thing to do.
The logic for 1.x remains unchanged, the semantics are the same, just written in a way that's easier to understand.

These are the particular HTTP versions: 0.9, 1.0, 1.1, 2.0, 3.0.
1.1 is supported by the SDK, 2.0/3.0 are currently not supported. We can make 1.0 work, in this case, regarding connection persistence defaults.

~There's no HTTP 1.1+, so hardening the version checks to pick the concrete protocol versions. And rejecting others (including the case which indicates a malformed response).~

### CODE REVIEW

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (RickWinter) [inline sdk/core/azure-core/src/http/curl/curl.cpp:958]: This comment doesn't make sense. You are throwing if its not 1.0 or 1.1 thus handling the 2.0 and 3.0 cases.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (RickWinter) [inline sdk/core/azure-core/src/http/curl/curl.cpp:959]: This should be greater than as in the original change. 1.2 can not break the functionality of 1.1,

Comment (RickWinter) [inline sdk/core/azure-core/src/http/curl/curl.cpp:959]:
```suggestion
    if (m_response->GetMajorVersion() == 1 && m_response->GetMinorVersion() >= 1)
```

**Review by RickWinter — DISMISSED** [review]

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (LarryOsterman) [inline sdk/core/azure-core/src/http/curl/curl.cpp:967]: Note that [HTTP 1.0](https://www.rfc-editor.org/rfc/rfc1945) does *NOT* define a Connection header (I checked the HTTP RFC last night). So there should be no circumstance in which we set m_httpKeepAlive for an HTTP 1.0 response.

If an HTTP 1.0 server response is received with a Connection header, the header should be ignored.

**Review by LarryOsterman — CHANGES_REQUESTED** [review]

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:959]: Is there an HTTP version 1.2? I don't see anything about this, we have 1.1 and then 2.0:
https://en.wikipedia.org/wiki/HTTP#Summary_of_HTTP_milestone_versions

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:967]: I am confused, are we sure that's the behavior we want? I thought in HTTP 1.0, an explicit presence of the keep-alive header is the opt-in for connection persistence.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection#directives
> Indicates that either the client or the server would like to close the connection. This is the default on HTTP/1.0 requests.

To bolster your finding, this does mention the following (citing the HTTP spec):
https://en.wikipedia.org/wiki/HTTP_persistent_connection#HTTP_1.0
> Under HTTP 1.0, connections should always be closed by the server after sending the response.

But also says:
```text
Since at least late 1995, developers of popular products (browsers, web servers, etc.) using HTTP/1.0, started to add an unofficial extension (to the protocol) named "keep-alive" in order to allow the reuse of a connection for multiple requests/responses.
...
When the server receives this request and generates a response, if it supports keep-alive then it also adds the same above header to the response. Following this, the connection is not dropped, but is instead kept open.
```

If what you described Larray is what we want as the behavior, I can certainly make the change to adhere to the spec. Do please confirm/let me know.

Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:965]: Discussed offline, we are OK with honoring the presence of the `Connection: keep-alive` header, rather than always setting `m_httpKeepAlive = false` for HTTP/1.0.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl.cpp:972]: Review note from @antkmsft : Change log message to HTTP/1.1x is supported rather than specifically 1.0/1.1.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl.cpp:977]: For reviewers: This is the primary change here that impacts the semantics. We are marking `httpKeepAlive` as false instead of true when we hit a case that is **not** HTTP 1.x, which is not supported by the SDK anyway.

Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:977]: @Jinming-Hu can you please review this PR, as an extra set of eyes. It doesn't change any supported scenarios for storage (since we only care about HTTP/1.1 there), and I don't expect it to make any perf impact.

I am requesting your review purely out of abundance of caution :)

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (antkmsft) [inline sdk/core/azure-core/src/http/curl/curl.cpp:972]: Instead of this message, let's update logging policy to log HTTP version for responses, and, optionally, for requests. The `Connection` header is already in the `g_defaultAllowedHttpHeaders`, so it should be sufficient info for diagnistic purposes - everything in a single log entry, plus it will cover both WinHTTP and libcurl adapters.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (antkmsft) [inline sdk/core/azure-core/src/http/curl/curl.cpp:952]: This comment section is not so much about what is happening in the code below, but it does describe what the specs say, so I see no reason to drop the portion about `HTTP/2` and `/3`.
```suggestion
    // (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection)
    // Should it come to HTTP/2 and HTTP/3, they are "keep-alive", but any response from HTTP/2 or
    // /3 containing a "Connection" header should be considered malformed. (HTTP/2:
    // https://httpwg.org/specs/rfc9113.html#ConnectionSpecific
    //  HTTP/3: https://httpwg.org/specs/rfc9114.html#rfc.section.4.2)
    //
    // HTTP/2+ are supposed to create persistent ("keep-alive") connections per host,
    // and close them by inactivity timeout. Given that we don't have such mechanism implemented
    // at this moment, we are closing all non-1.x connections immediately, which, in the worst case,
    // would only mean there's a perf hit, but the communication flow is expected to be correct.
```

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:972]: It's marked as verbose, I don't see any harm in keeping the log message here, as something exceptional has happened, even if it is specific to curl.

> Let's update logging policy to log HTTP version for responses, and, optionally, for requests

Can you please file an issue for that. It seems reasonable to do, imo, but not as part of this change.

Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/curl/curl.cpp:972]: FWIW, HTTP Requests *must* always be HTTP 1.1, as there are a more than trivial number of Azure services which will reject request that aren't on HTTP 1.1 :).

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (antkmsft) [inline sdk/core/azure-core/src/http/curl/curl.cpp:964]: Move tis logic from the above lines into HTTP/1 case:
```suggestion
      std::string connectionHeaderValue;
      {
        const Core::CaseInsensitiveMap& responseHeaders = m_response->GetHeaders();
        const auto connectionHeader = responseHeaders.find("Connection");

        if (connectionHeader != responseHeaders.cend())
        {
          connectionHeaderValue = Core::_internal::StringExtensions::ToLower(connectionHeader->second);
        }
      }

      const bool hasConnectionKeepAlive = connectionHeaderValue.find("keep-alive") != std::string::npos;

      if (m_response->GetMinorVersion() >= 1)
      {
        // HTTP/1.1+
        const bool hasConnectionClose = connectionHeaderValue.find("close") != std::string::npos;
        m_httpKeepAlive = (!hasConnectionClose || hasConnectionKeepAlive);
      }
      else
      {
        // HTTP/1.0
        m_httpKeepAlive = hasConnectionKeepAlive;
      }
```

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/http/curl/curl.cpp:972]: I know, but logging policy doesn't have to know that our transports do hardcode 1.1 for requests :)

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:964]: I would prefer to keep this PR scoped to the very specific change here that I proposed, so I can get it in.

Would you mind opening a PR for any other general refactoring/code clean-up that you think are good, **after** this one is merged? What you are suggesting seems reasonable.

**Review by antkmsft — COMMENTED** [review]
I don't understand why can't we change the scope of variables in the same PR. Before, they needed to have greater scope with the previous `if` arrangement. In this PR, we find it to be essential to rearrange `if` statements, and even change the logic, yet we don't think that moving the related variables closer into the scope meets the bar.

The same with the log message. We do think it is essential to put it, yet we don't care enough to put it in the logging policy which would cover both libcurl and WinHTTP cases, and the log message here will no longer be necessary, and it will only be adding additional noise to the logs. In the logging policy, we log HTTP verb, we log URL, we log headers, so why not log HTTP version?

**Comment by ahsonkhan** [comment]
At this time, I don't want to grow the scope/intent of this PR to what I have initially proposed (nor touch other lines of code). It has been waiting for approval for a week now, and it doesn't merit discussions on other changes (even if you consider them small).

If there's no blocking feedback on correctness, I would like to get this checked-in as is.

**Comment by antkmsft** [comment]
Here I have the change that I am talking about, you can see it is in the same scope and does not increase it, and it is not much more code: https://github.com/ahsonkhan/azure-sdk-for-cpp/pull/2

**Comment by RickWinter** [comment]
> Here I have the change that I am talking about, you can see it is in the same scope and does not increase it, and it is not much more code: [ahsonkhan#2](https://github.com/ahsonkhan/azure-sdk-for-cpp/pull/2)

I like moving the parsing to where its needed. Would bring a slight improvement to perf since in the 1.1 we don't need to parse both headers.

The clean output of the httpVersion in the log is really nice bonus :)

**Review by antkmsft — APPROVED** [review]
Ok, let's merge it, and I can create my PR on top of `main` instead of your PR - doesn't really matter that much how the change gets there.

**Review by Jinming-Hu — APPROVED** [review]

**Pull request merged**
